### PR TITLE
[FEATURE] Migrate FE hooks to PSR-15 middleware

### PR DESCRIPTION
**Pull request by bmack**
Used hooks in TSFE are deprecated, and PSR-15 middlewares and PSR-7 Request/Response data allows
for sending and evaluating crawler instructions in one go.

ToDo:
- We need to check "_CLI_crawler" and the ominuous Backend user Auth which is probably unneeded.
- We can probably just initialize BE_USER in the same middleware.

### CODE REVIEW

**Comment by bmack** [comment]
This one isn't tested yet, so let's wait a bit please.

File: Classes/Domain/Repository/QueueRepository.php [hunk 1]
```diff
@@ -511,4 +511,17 @@ public function getAvailableSets(): array
...
         return $rows;
     }
+
+    public function findByQueueId(string $queueId): ?array
+    {
+        $queueRec = $this->queryBuilder
+            ->select('*')
+            ->from($this->tableName)
+            ->where(
+                $this->queryBuilder->expr()->eq('qid', $this->queryBuilder->createNamedParameter($queueId))
```
Comment (tomasnorre) [inline Classes/Domain/Repository/QueueRepository.php]: shouldn't we add `\PDO::PARAM_STR` I'm not 100% when it's recommend and why not.

File: Classes/Middleware/CrawlerInitialization.php [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+     */
+    protected $queueRepository;
+
+    /**
+     * @param Context|null $context
+     */
+    public function __construct(Context $context = null)
+    {
+        $this->context = $context ?? GeneralUtility::makeInstance(Context::class);
+        $this->queueRepository = GeneralUtility::makeInstance(QueueRepository::class);
```
Comment (tomasnorre) [inline Classes/Middleware/CrawlerInitialization.php]: Why use the GeneralUtility and not the ObjectManager what is the win of this way, and when is GeneralUtility preferred over ObjectManager and vice versa.

File: Classes/Middleware/CrawlerInitialization.php [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        ];
+
+        $this->runPollSuccessHooks();
+
+        // Output log data for crawler (serialized content):
+        $content = serialize($GLOBALS['TSFE']->applicationData['tx_crawler']);
+        return new Response($content, 200);
+    }
+
+    protected function isRequestHashMatchingQueueRecord(?array $queueRec, string $hash): bool
```
Comment (tomasnorre) [inline Classes/Middleware/CrawlerInitialization.php]: Is the `?` here correct, don't recall seeing that notation before.

Comment (bmack) [reply Classes/Middleware/CrawlerInitialization.php]: yup. meaning array or null. PHP 7.1 feature. We return "null" from the QueueRepository that's why it is checked here.

File: Classes/Middleware/CrawlerInitialization.php (hunk 2 above)
Comment (bmack) [reply Classes/Middleware/CrawlerInitialization.php]: If you have inject* methods you need to use ObjectManager. But it adheres a dependency on Extbase, so GeneralUtility is sufficient in most cases.

In some cases ObjectManager needs TypoScript configuration to fetch the right objects (in Extbase plugins mostly), which you see this more in third-party-extensions based on Extbase,

File: Classes/Domain/Repository/QueueRepository.php (hunk 1 above)
Comment (bmack) [reply Classes/Domain/Repository/QueueRepository.php]: `\PDO::PARAM_STR` is the default argument.

File: Classes/Middleware/CrawlerInitialization.php [hunk 4]
```diff
@@ -57,7 +58,7 @@ class CrawlerInitialization implements MiddlewareInterface
     public function __construct(Context $context = null)
     {
         $this->context = $context ?? GeneralUtility::makeInstance(Context::class);
```
Comment (tomasnorre) [inline Classes/Middleware/CrawlerInitialization.php]: This is also GeneralUtility

File: Classes/Domain/Repository/QueueRepository.php (hunk 1 above)
Comment (tomasnorre) [reply Classes/Domain/Repository/QueueRepository.php]: Didn't know that.

**Comment by bmack** [comment]
Works fine, all tested now.

**Pull request merged**